### PR TITLE
feat: relay offline sync notice/detection

### PR DESCRIPTION
**Pull request by pvinis**
This PR resolves [MOPLAT-546]

### Description

We generate a checksum from the files that are involved with the offline sync (and their deps, like fragments). this way we know if we need to resync.

We have not done anything "smart" about removals, so if we remove something like `title` from a gql query, this would say we should resync, when we don't really need, but its a general rule here that if we update something, we should resync.

### PR Checklist

- [ ] I tested my changes on **iOS** and **Android**.
- [ ] I added screenshots or videos to illustrate my changes.
- [ ] I added Tests and Stories for my changes.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to run this PR on the simulator or device.

Need help with something? Have a look at our [docs], or get in touch with us.

[docs]: /docs/README.md

[MOPLAT-546]: https://artsyproduct.atlassian.net/browse/MOPLAT-546?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "prettier-write": "prettier --write --ignore-unknown",
     "relay:watch": "yarn relay --watch",
     "relay": "yarn relay-compiler",
+    "postrelay": "./scripts/relay-checksum.mjs",
```
Comment (pvinis) [inline package.json:19]: just a little postscript, in case we run this manually.

File: src/app/system/store/Models/GlobalStoreModel.ts [hunk 2]
```diff
@@ -10,11 +10,7 @@ import { getNetworkStatusModel, NetworkStatusModel } from "./NetworkStatusModel"
 import { getPresentationModeModel, PresentationModeModel } from "./PresentationModeModel"
 import { SelectModeModel, getSelectModeModel } from "./SelectModeModel"
 
-type ActiveMode = "viewer" | "manager"
-
 interface GlobalStoreStateModel {
-  activeMode: ActiveMode
```
Comment (pvinis) [inline src/app/system/store/Models/GlobalStoreModel.ts:16]: this was not used

File: src/app/system/store/Models/GlobalStoreModel.ts [hunk 3]
```diff
@@ -10,11 +10,7 @@ import { getNetworkStatusModel, NetworkStatusModel } from "./NetworkStatusModel"
...
-  activePartnerID: string | null
```
Comment (pvinis) [inline src/app/system/store/Models/GlobalStoreModel.ts:17]: moved this below, for nicer separation

File: src/app/system/store/Models/GlobalStoreModel.ts [hunk 4]
```diff
@@ -27,8 +23,11 @@ interface GlobalStoreStateModel {
 }
 
 export interface GlobalStoreModel extends GlobalStoreStateModel {
+  activePartnerID: string | null
   setActivePartnerID: Action<this, string | null>
-  setActiveMode: Action<this, ActiveMode>
+
+  latestSyncedChecksum: string
```
Comment (pvinis) [inline src/app/system/store/Models/GlobalStoreModel.ts:29]: !!

File: src/app/system/sync/currentRelayChecksum.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+/**
+ * THIS FILE IS AUTOGENERATED. DO NOT EDIT MANUALLY.
+ *
+ * This checksum is generated from all the files in the `__generated__` directory that are
+ * also part of the `syncManager` file, and any dependency they might have (eg. fragments).
+ */
+
+export const currentRelayChecksum = "9f0a70d6b486138abe4e7e754e7c938898e52b3b"
```
Comment (pvinis) [inline src/app/system/sync/currentRelayChecksum.ts:8]: !!

File: scripts/relay-checksum.mjs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  for (const [index, item] of items.entries()) {
+    data.push(await callback(item, index))
+  }
+  return data
+}
+
+const calculateChecksum = async () => {
+  const content = fs.readFileSync(QUERIES_FILE, "utf-8")
+
+  const gqlFiles = content.match(/(__generated__\/.*)/g)
```
Comment (pvinis) [inline scripts/relay-checksum.mjs:23]: no need to do anything for fragments, as they are included in the query files. if a fragment changes, the query file it's used in also changes.

File: .husky/pre-push [hunk 7]
```diff
@@ -1,4 +1,6 @@
 #!/usr/bin/env sh
 . "$(dirname -- "$0")/_/husky.sh"
 
+
 yarn type-check
+yarn relay && git add ./src/app/system/sync/currentRelayChecksum.ts
```
Comment (damassi) [inline .husky/pre-push:6]: Can we move this to `/data` folder alongside our other transient graphql schema data?

File: scripts/relay-checksum.mjs [hunk 8]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env node
+
+// @ts-check
+
+import fs from "fs"
+import sha1sum from "sha1-sum"
+import sha1 from "sha1"
+
+const QUERIES_FILE = "src/app/system/sync/syncManager.ts"
+const CHECKSUM_FILE = "src/app/system/sync/currentRelayChecksum.ts"
```
Comment (damassi) [inline scripts/relay-checksum.mjs:10]: `/data/currentRelayChecksum`

File: scripts/relay-checksum.mjs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  if (!gqlFiles) {
+    throw new Error("No graphql files found. Are you pointing to the right QUERIES_FILE?")
+  }
+
+  const paths = gqlFiles.map((file) => `src/${file.slice(0, -1)}.ts`)
+
+  const checksums = await mapAsyncLinear(paths, async (path) => await sha1sum(path))
+  const finalChecksum = sha1(checksums.join("\n"))
+
+  // console.log(finalChecksum)
```
Comment (damassi) [inline scripts/relay-checksum.mjs:33]: 🦑

File: src/app/screens/Settings/OfflineModeSettings.tsx [hunk 10]
```diff
@@ -18,6 +19,8 @@ export const OfflineModeSettings = () => {
   const { relayEnvironment } = useSystemRelayEnvironment()
 
   const partnerID = GlobalStore.useAppState((state) => state.activePartnerID)!
+  const latestSyncedChecksum = GlobalStore.useAppState((state) => state.latestSyncedChecksum)!
```
Comment (damassi) [inline src/app/screens/Settings/OfflineModeSettings.tsx:22]: Maybe just `syncedChecksum`, or better: `offlineChecksum`? Then a future dev can understand that this is very specifically related to offline mode and doesn't need to hunt. Also something I notice often (which is a bit confusing) is a lot of reliance on `latest`, `current`, `final` interstitial naming states in prepended to variable names. `syncedChecksum` describes things perfectly

File: src/app/system/sync/currentRelayChecksum.ts (hunk 5 above)
Comment (damassi) [reply src/app/system/sync/currentRelayChecksum.ts:8]: `/data`

**Review by damassi — APPROVED** [review]
Looks amazing @pvinis 💯, nice and simple. A couple suggestions about file locations and naming but otherwise 👍 👍

**Comment by pvinis** [comment]
adding #sqongr after chat with chris

**Pull request merged**